### ISSUE TITLE
RFC: regex-impl

### ISSUE THREAD

**Issue by nochiel** [mawww/kakoune#1660]
Re. commit 065bbc8f5903c39a7ed4cca60d721dc92741154a

I was wondering why it was decided that a bespoke regex implementation was more desirable than using a library such as [RE2](https://github.com/google/re2) ?  (If this has already been documented please point me to the discussion.)

(p.s. I recall that there was recently a similar discussion on [emacs-devel](http://lists.gnu.org/archive/html/emacs-devel/2016-12/msg00622.html) about emacs' custom regex vs the difficulty of using a library.)

**Comment by lenormf** [comment]
```
2017-10-20 12:24:15     Franciman       and did you see re2?
2017-10-20 12:24:48     mawww   I did, but it does not support arbitrary iterators, it wants const char* everywhere
2017-10-20 12:24:53     Franciman       Ah cool
2017-10-20 12:25:14     Franciman       I asked because last time I checked they used some optimization stuff for the VM
2017-10-20 12:25:17     mikaoP  mawww: why do you think that is bad use boost-regex?
2017-10-20 12:25:36     Franciman       so maybe one could get some speed. But not sure it's needed, though
2017-10-20 12:25:54     Franciman       and also if it's applicable, since many times regex are used once
2017-10-20 12:25:54     mawww   its a big dependency, and one big feature of the custom regex impl is that we can efficiently search backwards with it
2017-10-20 12:26:12     Franciman       cool!
2017-10-20 12:27:01     mawww   RE2 is a much bigger piece of code that the custom regex impl, its going to be much more efficient in many cases, but the custom impl is already as fast or faster as boost regex in most case, and should not  have any pathological cases
2017-10-20 12:28:12     Franciman       I see
2017-10-20 12:28:16     mawww   so, as with RE2, we should not see any regex taking huge amount of time, we have a O(N*M) complexity with N the number of instructions in the regex (roughly proportional to the regex length in char) and M th
e length of the subject string
2017-10-20 12:28:58     Franciman       boost-regex used backtracking?
2017-10-20 12:29:08     Franciman       uses*
2017-10-20 12:29:18     mawww   but as the implementation is much simpler than RE2, we are not as fast as it can get, for example in RE2 regex like ".{10}foobar" are very fast, because re2 will be able to detect it can just search for the  literal foobar and then replay from 10 character before
```

**Comment by mawww** [comment]
So, in more compact form:

* We need an engine that can match on non contiguous text.
* We want more features than whats offered by std::regex (in particular we want `\K`, `\A`, `\z`, lookbehinds...)
* Boost is a huge dependency
* Efficient backward matching is an added benefit

No existing regex engine provides support for all that, so I ended up writing a new one, specially tailored for Kakoune's use case.

Another reason is that it was a fun enough project to keep me interested in working on that, I learned a lot. And its still reasonably compact (<2000 lines).
